crates/net/network/src/session/mod.rs

```diff
@@ -170,9 +171,63 @@ impl SessionManager {
         self.status
     }
 
-    /// Returns the session hello message.
-    pub fn hello_message(&self) -> HelloMessageWithProtocols {
-        self.hello_message.clone()
+    /// Returns the session default hello message.
+    pub fn hello_message(&self) -> &HelloMessageWithProtocols {
+        &self.hello_message
+    }
+
+    // Update hello message with extra protocols to support in the session.
+    fn update_hello_message(
+        &self,
+        extra_conns: &[Box<dyn DynConnectionHandler>],
+    ) -> HelloMessageWithProtocols {
+        let hello_message = self.hello_message.clone();
+        if extra_conns.is_empty() {
+            return hello_message
+        }
+
+        let mut builder = HelloMessageBuilder::new_from(hello_message);
+        for handler in extra_conns {
+            builder = builder.protocol(handler.protocol());
+        }
+
+        builder.build()
+    }
+
+    /// Returns connection handlers for the extra protocols that should be supported in the
+    /// incoming session.
+    pub fn extra_protocols_on_incoming(
+        &self,
+        remote_addr: SocketAddr,
+    ) -> (HelloMessageWithProtocols, Vec<Box<dyn DynConnectionHandler>>) {
+        let mut extra_conns = Vec::with_capacity(self.extra_protocols.len());
+        for protocol in self.extra_protocols.iter() {
+            if let Some(handler) = protocol.on_incoming(remote_addr) {
+                extra_conns.push(handler);
+            }
+        }
+
+        let hello_message = self.update_hello_message(&extra_conns);
+
+        (hello_message, extra_conns)
+    }
+
+    /// Returns connection handlers for the extra protocols that should be supported in the
+    /// outgoing session.
+    pub fn extra_protocols_on_outgoing(
+        &self,
+        remote_addr: SocketAddr,
+        remote_peer_id: PeerId,
+    ) -> (HelloMessageWithProtocols, Vec<Box<dyn DynConnectionHandler>>) {
```

Comment:
this is correct, we want only include protocols depending on the direction

though I think this can be simplified by filtering first and then pushing the extra protocols right before we we authenticate in `authenticate`

ref https://github.com/paradigmxyz/reth/pull/5559

Reply:
I assumed that would happen as part of methods `on_incoming` and `on_outgoing` on protocol handler and return connection handler as one atomic operation. it's cheaper than first cloning the handlers, the conns returned can be moved into the spawned thread.

Reply:
It does happen the way I assumed, didn't catch that you put `DynConnectionHandler`s into a struct called `RlpxSubProtocolHandlers`. anyhow, would have needed `Arc<DynProtocolHandler>` to clone which isn't expensive.